packages/adp-tooling/src/types.ts

```diff
@@ -138,3 +142,119 @@ export const enum HttpStatusCodes {
     NOT_IMPLEMETED = 501,
     SERVICE_UNAVAILABLE = 503
 }
+
+export type Writer = new (fs: Editor, projectPath: string) => IWriter<any>;
+export type IWriterData<T extends ChangeType> = IWriter<GeneratorData<T>>;
+export interface IWriter<T> {
+    /**
+     * Writer function.
+     *
+     * @param {T} data - Data needed for the writer function.
+     */
+    write(data: T): Promise<void>;
+}
+
+export const enum ChangeType {
+    ADD_NEW_MODEL = 'appdescr_ui5_addNewModel',
+    ADD_ANNOTATIONS_TO_ODATA = 'appdescr_app_addAnnotationsToOData',
+    CHANGE_DATA_SOURCE = 'appdescr_app_changeDataSource',
+    ADD_COMPONENT_USAGES = 'appdescr_ui5_addComponentUsages',
+    ADD_LIBRARY_REFERENCE = 'appdescr_ui5_addLibraries',
+    CHANGE_INBOUND = 'appdescr_app_changeInbound'
+}
+
+export type GeneratorData<T extends ChangeType> = T extends ChangeType.ADD_ANNOTATIONS_TO_ODATA
+    ? AnnotationsData
+    : T extends ChangeType.ADD_COMPONENT_USAGES
+    ? ComponentUsagesData
+    : T extends ChangeType.ADD_LIBRARY_REFERENCE
+    ? ComponentUsagesData
+    : T extends ChangeType.ADD_NEW_MODEL
+    ? NewModelData
+    : T extends ChangeType.CHANGE_DATA_SOURCE
+    ? DataSourceData
+    : T extends ChangeType.CHANGE_INBOUND
+    ? InboundData
+    : never;
+
+export interface AnnotationsData {
+    projectData: AdpProjectData;
+    timestamp: number;
+    annotationFileName?: string;
+    isInternalUsage: boolean;
+    oDataSource: string;
+    annotationFilePath: string;
+}
+export interface ComponentUsagesData {
```

Comment:
Could you please consider structuring this a bit better instead of keeping it flat? I see two general properties, then a bunch for a component and and a bunch for a possible library.

Reply:
Changed in: [706928d](https://github.com/SAP/open-ux-tools/pull/1643/commits/706928d0e145dee24ff454510c00cb9832238bec)